core/src/main/java/org/apache/iceberg/io/FanoutFileWriter.java

```diff
@@ -0,0 +1,107 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.Map;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.StructLikeMap;
+
+/**
+ * A writer capable of writing to multiple specs and partitions that keeps files for each
+ * seen spec/partition pair open until this writer is closed.
+ * <p>
+ * As opposed to {@link ClusteredFileWriter}, this writer does not requite the incoming records
+ * to be clustered by partition spec and partition. The downside is potentially larger memory
+ * consumption.
+ */
+abstract class FanoutFileWriter<T, R> implements PartitionAwareFileWriter<T, R> {
+
+  private final Map<Integer, Map<StructLike, FileWriter<T, R>>> writers = Maps.newHashMap();
+  private boolean closed = false;
+
+  protected abstract FileWriter<T, R> newWriter(PartitionSpec spec, StructLike partition);
+
+  protected abstract void addResult(R result);
+
+  protected abstract R aggregatedResult();
+
+  public CharSequence currentPath(PartitionSpec spec, StructLike partition) {
+    return ((RollingFileWriter<?, ?, ?>) writer(spec, partition)).currentPath();
+  }
+
+  public long currentPosition(PartitionSpec spec, StructLike partition) {
+    return ((RollingFileWriter<?, ?, ?>) writer(spec, partition)).currentRows();
+  }
+
+  @Override
+  public void write(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    FileWriter<T, R> writer = writer(spec, partition);
+    writer.write(row);
+  }
+
+  private FileWriter<T, R> writer(PartitionSpec spec, StructLike partition) {
+    Map<StructLike, FileWriter<T, R>> specWriters = writers.computeIfAbsent(
+        spec.specId(),
+        id -> StructLikeMap.create(spec.partitionType()));
+    FileWriter<T, R> writer = specWriters.get(partition);
+
+    if (writer == null) {
+      // copy the partition key as the key object is reused
+      StructLike copiedPartition = partition != null ? StructCopy.copy(partition) : null;
+      writer = newWriter(spec, copiedPartition);
+      specWriters.put(copiedPartition, writer);
+    }
+
+    return writer;
+  }
```

Comment:
Could we maintain those writers from different partition spec and different partition into a flatten Map<K, FileWriter<T,R>> rather than a nested Map,  by using a composited key <partitionSpecId, partitionData> .   I think that will make the writer choose strategy quite simple,  it will also simplify the ClusteredFileWriter's write method.

Reply:
I can change that. No preference from my side here.